distributed/worker_state_machine.py

```diff
@@ -1478,17 +1478,18 @@ def _purge_state(self, ts: TaskState) -> None:
         This is idempotent
         """
         logger.debug("Purge task: %s", ts)
-        key = ts.key
-        self.data.pop(key, None)
```

Comment:
It's not clear to me why a pop would unspill anything. Isn't this rather a big of the spill layer?

Reply:
Because pop() is (correctly) implemented as
```python
def pop(self, key, default=no_default):
    try:
        return self[key]
    except KeyError:
        if default is no_default:
            raise
        return default
```

Reply:
> d.pop(key, None) is a popular hack around the fact that dicts don't have a discard method. But the primary purpose of pop() is to return the key if it exists, not to quietly delete it if it exists.

I know. 

> Because pop() is (correctly) implemented everywhere as

I could just see this being implemented as

```python
def pop(self, key, default=no_default):
    if key in self:
        value = self[key]
        del self[key]
        return value
    elif default is no_default:
        raise
    return default
```

which would avoid unspilling. Implementing this in the spillbuffer seems more reasonable than trying to avoid `pop`. After all, with this limitation, pop should maybe not even be implemented on this mapping.

I don't have a strong preference, I just think "fixing" this in the spillbuffer is a more robust fix than avoiding pop

Reply:
It would not avoid unspilling. The point is that the key *is* in the SpillBuffer when the free-keys event triggers `_purge_state`.

Reply:
I agree that adding a tripwire in SpillBuffer.pop() is a good idea. Added one now.